### PR TITLE
Create `chainer.optimizer_hooks` namespace and move hooks there.

### PR DESCRIPTION
**Pull request by delta2323**
This PR is a updated version of #2397.

### CODE REVIEW

File: chainer/optimizer.py [hunk 1]
```diff
@@ -603,171 +595,42 @@ def __set__(self, obj, value):
...
+    deprecated = set(deprecated)
 
-    .. math::
+    class Wrapper(object):
 
-        g_t \\leftarrow g_t + N(0, \\sigma_t^2)
+        def __getattr__(self, attr):
+            if attr in deprecated:
+                warnings.warn(
+                    "chainer.optimizer.{0} is deprecated since v3. "
```
Comment (unnonouno) [inline chainer/optimizer.py]: Use single quotes

File: chainer/optimizer.py [hunk 2]
```diff
@@ -603,171 +595,42 @@ def __set__(self, obj, value):
...
+                    DeprecationWarning)
 
-    where
+            return getattr(mod, attr)
 
-    .. math::
+        def __setattr__(self, attr, value):
+            if attr in deprecated:
+                warnings.warn(
+                    "chainer.optimizer.{0} is deprecated since v3. "
```
Comment (unnonouno) [inline chainer/optimizer.py]: ditto

File: chainer/optimizer.py [hunk 3]
```diff
@@ -603,171 +595,42 @@ def __set__(self, obj, value):
...
-    def __init__(self, lower_bound, upper_bound):
-        self.lower_bound = lower_bound
-        self.upper_bound = upper_bound
-
-    def __call__(self, rule, param):
-        grad = param.grad
-        xp = cuda.get_array_module(grad)
-        with cuda.get_device_from_array(grad):
-            xp.clip(grad, self.lower_bound, self.upper_bound, out=grad)
+sys.modules[__name__] = deprecation_warning_wrapper(
```
Comment (unnonouno) [inline chainer/optimizer.py]: Is this a recommended method to show warnings? Do you have simpler method?

**Comment by delta2323** [comment]
Closed, as it is hard to resolve conflicts than to make a new PR (#3977)

**Pull request closed without merging**